### ISSUE TITLE
[HOLD for payment 2023-02-08] [$1000] Backend - Attachment links capture the environment they are from

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#14227]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Use https://staging.new.expensify.com/ to chat with somebody
2. Send an attachment
3. Use https://new.expensify.com/ to open the same conversation
4. Inspect the URL for the attachment
5. It will say it's loaded from https://staging.expensify.com/chat-attachments/ even though you are on new.expensify

## Expected Result:
all attachment URLs to be resolved relative to current environment
E.g. when I’m on staging.new I expect all attachments to be resolved from https://staging.new.expensify.com/chat-attachments/…
Or when Im on prod I expect all attachments to be resolved from https://new.expensify.com/chat-attachments/…

## Actual Result:
some attachment urls are resolved from [staging.expensify.com](http://staging.expensify.com/) others from [new.expensify.com](http://new.expensify.com/) 

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?

- [ ] MacOS / Chrome / Safari

**Version Number:** 1.2.52-4
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 
![image](https://user-images.githubusercontent.com/43996225/211927253-2ca18c54-f8a4-4a35-abfc-feef9441c85c.png)
<img width="960" alt="Untitled" src="https://user-images.githubusercontent.com/43996225/211927271-7b5a895f-19fb-4bf6-a99b-c019abe602ba.png">

**Expensify/Expensify Issue URL:**
**Issue reported by:** @kidroca
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1673289214225929

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01911d83399c4bca37</li>
        <li>Upwork Job ID: 1613477091071528960</li>
        <li>Last Price Increase: 2023-01-20</li>
    </ul>
</details>

**Comment by getusha** [comment]
This happens when you upload the attachment on `staging` and view it on `production` vice versa, and i don't think this is an issue.

cc @bfitzexpensify

**Comment by Prince-Mendiratta** [comment]
# Proposal
## RCA
When the attachment is uploaded, the HTML is sent to the backend using the staging URL. When the same is received via the API, the HTML remains same and the src is not updated.
![Screenshot_20230112_155211](https://user-images.githubusercontent.com/54077356/212041504-e43e2100-3644-4b1a-94e0-fe03bb2071fe.png)

## Solution
This should be handled on the backend, the file links should have the base URL updated depending on the request host.

**Comment by kidroca** [comment]
@Prince-Mendiratta 
> When the attachment is uploaded, the **HTML is sent to the backend** using the staging URL

Can you point where this is happening? I think the HTML is formed on the backend

If the `html` is formed locally I think it's should be ok to set the image source as 

```html
<img src="/chat-attachments/{path}" />
```

**Comment by mountiny** [comment]
Asking in Slack for clarification https://expensify.slack.com/archives/C049HHMV9SM/p1673539365623209?thread_ts=1673289214.225929&cid=C049HHMV9SM

**Comment by jatinsonijs** [comment]
I think it's ok to have same url for both staging and production as we are already sharing user data between staging and production. And if from backend we remove base url from image it will not work for Android and iOS. we have to explicitly set base url.

**Comment by kidroca** [comment]
> This happens when you upload the attachment on `staging` and view it on `production` vice versa, and i don't think this is an issue.

This is inconsistent behavior - if it should be possible to load `staging` attachments from `production` and `production` attachments from `staging`, why not simply always capture all attachments under `https://www.expensify.com` - at least all would be prefixed by the same origin URL

**Comment by getusha** [comment]
That means we have to send the picture both to `production` and `staging` to achieve this consistency?

**Comment by kidroca** [comment]
> That means we have to send the picture both to `production` and `staging` to achieve this consistency?

We don't need to send to both, if staging can access attachments from `www.expensify...` and prod can access attachments from the same url, then we can at least only use that one

---

I think attachments are stored in the same place and can be accessed with different urls, like I've pointed out here: https://expensify.slack.com/archives/C049HHMV9SM/p1673541181055489?thread_ts=1673289214.225929&cid=C049HHMV9SM

You can open the sam attachment by changing `www.expensify...` to `staging.expensify...`

**Comment by kidroca** [comment]
For more context, we're trying to change the way we load attachments and start to use an authentication header 
HTML images (src attribute) do not support headers, so to load and image we'll have to first make a `fetch` request which is subject to CORS 

When we are on `new.expensify.com` but try to load something from `staging.expensify.com/...` a CORS request will be made. Image elements handle this well, but `fetch` would not expose the underlying `blob` unless the server responds in a CORS friendly way

So an alternative might be to configure CORS and allow requests from new.staging to www.expensify and the various combinations:
- new.expensify -> www.expensify
- new.expensify -> staging.expensify
- staging.new. -> www.expensify
- staging.new. -> staging.expensify

And keep in mind to update this if in the future attachments can be saved under `new.expensify` 

I think it would be simpler to save the attachment and excluding the origin and allow to load it from all these domains 
web would use the current origin, while native apps can use the origin stored in App Config

**Comment by kidroca** [comment]
After further investigation I found logic inside App that expects attachment URLs to be coming from `Config.EXPENSIFY.EXPENSIFY_URL` and replace them to be loaded from `Config.EXPENSIFY.URL_API_ROOT` 

https://github.com/Expensify/App/blob/81855c2dbcdd2e95bb7d86ec96d1857241ae2a6f/src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js#L38-L45

So it seems the intended way to render user attachments (that were uploaded to Expensify) it so load them from `EXPENSIFY.URL_API_ROOT` 

Of course `EXPENSIFY.EXPENSIFY_URL` is a constant so it can only match `staging` or `prod` so if the attachment urls are stuck with the host they were made on this logic would not always work (and some attachments would not be loaded from `URL_API_ROOT` when they should have)

---

There seems to be 3 ways to solve this

### 1. Handle multiple expensify domains and rewrite the source
(No backend changes - handle more than one origins on the client side)

```js
/**
 * Update the image URL so images can be accessed depending on the config environment
 *
 * @param {String} urlString
 * @returns {*|String}
 */
function mapSource(urlString) {
    const url = new URL(urlString);
    if (/expensify/i.test(url.origin)) {
        return `${Config.EXPENSIFY.URL_API_ROOT}${url.pathname.slice(1)}`;
    }

    return urlString;
}

const previewSource = mapSource(htmlAttribs.src);
const source = mapSource(isAttachment
    ? htmlAttribs[CONST.ATTACHMENT_SOURCE_ATTRIBUTE]
    : htmlAttribs.src);
```

### 2. Always save attachments to `www.exepnsify.com` 
Update backend and the logic here to always expect attachments from this URL
This would keep almost the same logic for `ImageRenderer` sources

### 3. Let the backend sort this out
I think it might be best to just handle this on the backend and drop the client logic

Backend flow:
1. Upload attachment request arrives
2. Attachment is saved to storage
3. Absolute path (excluding host) is saved to database (e.g. `/chat-attachments/123.../...`)

Then when report actions are returned the backend can return them with the abs path
(like originally proposed here: https://github.com/Expensify/App/issues/14227#issuecomment-1380311610)

**Comment by mountiny** [comment]
thanks @kidroca 

continuing the discussion in [slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1673877190689849) as I would love to get more people to agree with this change before updating anything

**Comment by kidroca** [comment]
Discovered a related problem regarding PDF document previews

If a PDF was uploaded on staging and later viewed on prod, clicking on the attachment would result in an Error

https://user-images.githubusercontent.com/12156624/212701874-666c1c3c-e26e-4406-9331-feeafb99b525.mp4

**Comment by mountiny** [comment]
@kidroca I think we have internally agreed that saving and using the absolute path is the best way forwards. It will require some back end changes, but I wont get to them until next week. However, we should make sure to update the client to handle the absolute path, as there will be both, full urls and absolute paths being sent for some time possibly.

Would you be able to propose changes to client to handle both (check if the path starts with `/` and treat it as absolute path) and whatever we have now? I think we would then filter this in API layer to only respond with the absolute path once this is finished.

**Comment by kidroca** [comment]
@mountiny 
> Would you be able to propose changes to client to handle both (check if the path starts with `/` and treat it as absolute path) and whatever we have now? I think we would then filter this in API layer to only respond with the absolute path once this is finished.

Yes, but I'll need some details

When we have an absolute path (url starts with `/`) should it be resolved using `URL_API_ROOT` or we can let it use the current origin?

When I'm on `new.expensify.com` - `URL_API_ROOT` is `www.expensify.com` 
When I'm on `staging.new.expensify.com` - `URL_API_ROOT` is `staging.expensify.com` 

So if we don't (have logic to) prefix absolute urls with `URL_API_ROOT` they will trigger a request against `{origin}/{path}` and not `{URL_API_ROOT}/{path}` 

Would the backend be updated is such a way to just allow `{origin}/{path}` - no client side logic 
Or we'll still have to have small client side logic to create `{URL_API_ROOT}/{path}`

**Comment by mountiny** [comment]
@kidroca I think we would do `URL_API_ROOT`, the data is still served from `www.expensify.com`

@bfitzexpensify can you manually update the price in upwork back to $1000? i just forgot to hire Kidroca last week.

**Comment by mountiny** [comment]
Ok, slay, @thienlnam is king, no need to do it manually

**Comment by kidroca** [comment]
> @kidroca I think we would do `URL_API_ROOT`, the data is still served from `www.expensify.com`

I think this might still need some CORS settings update
Once we try to fetch an Image with auth header - browsers would expect the backend to respond that the given header is allowed ([Access-Control-Allow-Headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Headers))

**Comment by kidroca** [comment]
#### My proposal would be to have minimal client logic, ebd result being this

Replace existing logic with a check for absolute path (`/`) and prefix with `URL_API_ROOT` 

### Though it would have to happen in stages

#### Stage 1
1. Add logic that tests for absolute path `URL_API_ROOT` 
2. Add/update logic that tests whether the given URL is for staging or prod and replace with `URL_API_ROOT` 

#### Stage 2
Run a DB migration that strips the origin when it is staging or prod

#### Stage 3 
Now step 2) from Stage 1 becomes unnecessary as there should be no attachments with a url from staging or prod - we can remove the logic

---

Whatever is decided, we should at least do the Stage 1 changes, the rest is optional / nice to have

**Comment by kidroca** [comment]
PR ready for review

**Comment by mountiny** [comment]
Thank you! Will have a look!

**Comment by kidroca** [comment]
I think I've found a CORS problem loading images with headers 

Trying to load `https://www.expensify.com/chat-attachments/...` from `new.expensify.com` would result in this error

```
[Error] Preflight response is not successful. Status code: 302
[Error] Fetch API cannot load https://www.expensify.com/chat-attachments/4198796077944938521/w_6f21f50b3af2ee93c4e6e288ba5b986858166367.jpg.1024.jpg due to access control checks.
[Error] Failed to load resource: Preflight response is not successful. Status code: 302 (w_6f21f50b3af2ee93c4e6e288ba5b986858166367.jpg.1024.jpg, line 0)
[Error] Preflight response is not successful. Status code: 302
```
- _I was able to simulate this in Safari, by resolving `new.expensify.com` to `localhost`_ 

Because we make a "complex" request (due to having a custom header like `X-Chat-Attachment-Token`)
browsers would make a **preflight** OPTIONS request and only continue with the actual request if access 
rules are satisfied (from OPTIONS response)

I think the backend doesn't expect and handle this OPTIONS request (for attachments) as of now, because 
if we make an options requests against an attachment path we don't get an OK 200 response but a 302

```sh
curl -X OPTIONS \
https://www.expensify.com/chat-attachments/4198796077944938521/w_6f21f50b3af2ee93c4e6e288ba5b986858166367.jpg.1024.jpg \
--dump-header -
```

<details>
<summary>Results in</summary>

```
HTTP/2 302
date: Fri, 27 Jan 2023 16:54:06 GMT
content-type: text/html; charset=UTF-8
location: https://www.expensify.com?exitTo=%2Fchat-attachments%2F4198796077944938521%2Fw_6f21f50b3af2ee93c4e6e288ba5b986858166367.jpg.1024.jpg
set-cookie: initialReferer=%28direct%29; expires=Mon, 24-Jan-2033 16:54:06 GMT; Max-Age=315360000; path=/; domain=.expensify.com; secure; HttpOnly
vary: Accept-Encoding
x-xss-protection: 1; mode=block
strict-transport-security: max-age=31536000; includeSubDomains; preload
x-content-type-options: nosniff
referrer-policy: origin-when-cross-origin
content-security-policy: frame-ancestors https://we.are.expensify.com www.expensify.com https://viewer.expensify.com chrome-extension://oiicpdkmeclmgmlmbajefnkalcfageek chrome-extension://eeibfofdhodcjojmbookcpigoaeilkek https://new.expensify.com
cache-control: no-cache, no-store
cf-cache-status: DYNAMIC
set-cookie: browserGUID=63d401ae6e6b2; expires=Mon, 24-Jan-2033 16:54:06 GMT; Max-Age=315360000; path=/; domain=.expensify.com; secure; HttpOnly
set-cookie: __cf_bm=YcI7lvHtJeVSdv.HTXbcxpXQ8aR4CLRngcgbx09QYP4-1674838446-0-AQB0Dgxji3m8IZmce2ngZB1qfUfO6w9pEvb7fUgElkDBp66yMR5mLCICmA5R2zyvbmsL3gilypj3JtL908qpFsc=; path=/; expires=Fri, 27-Jan-23 17:24:06 GMT; domain=.expensify.com; HttpOnly; Secure; SameSite=None
set-cookie: _cfuvid=uCAzCY9taFvkhdf53KWVPNTmqWnJe1_EYu_yCEH2NPQ-1674838446570-0-604800000; path=/; domain=.expensify.com; HttpOnly; Secure; SameSite=None
server: cloudflare
cf-ray: 790302217b870c6f-SOF
```
</details>

There's no way to opt out of the preflight request - it's a security standard for web browsers
The preflight request is considered successful if it resolves with a status code of 200 
and when access rules like [Access-Control-Allow-Headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Headers) are satisfied by the response

**Comment by kidroca** [comment]
I think the simplest fix would be to just make requests against the current origin and resolve the path internally 

So if I'm on `new.expensify` I make request against `new.expensify.com/chat-attachments/{path}` - the browser does not run any access control checks because the request is for the same origin
The server that hosts `new.expensify` would have to know how to resolve the path `new.expensify.com/chat-attachments/{path}` and serve the attachment

Alternatively the backend can be updated to respond to the OPTIONS request described in [my prior comment](https://github.com/Expensify/App/issues/14227#issuecomment-1406800255)
But I think we're introducing CORS checks artificially, because to us `www.expensify` and `new.expensify` are the same origin, while for browsers they are different origins and are subject to CORS checks

**Comment by mountiny** [comment]
Thank you! Asking internally, I think the solution makes sense but confirming what it would entail.

**Comment by mountiny** [comment]
This might be a bit more tricky, basically the API is hosted on the `expensify.com/api.php?command` route and it seems like it would not be that easy to "funnel" the response through the `new.expensify.com`  to avoid cors.

**Comment by mountiny** [comment]
Can you remind me, what would be the trigger point of running into these CORS issues? I thought we have already started adding the custom header token `X-Chat-Attachment-Token`. Thank you!

**Comment by kidroca** [comment]
> Can you remind me, what would be the trigger point of running into these CORS issues? I thought we have already started adding the custom header token `X-Chat-Attachment-Token`. Thank you!

Correct, the "trigger point" would be the request with header

We're already using the custom header in native iOS/Android apps - no problems there, because they aren't subject to CORS checks (they can make direct requests to any URL) - only browsers run CORS checks

In the past (no header) - there were fewer limitations, it was possible to load images from different origins without CORS checks.
With headers, the request is subject to CORS

But the request is only subject to CORS because of the slight origin mismatch `new.expensify` vs `www.expensify` 

> This might be a bit more tricky, basically the API is hosted on the `expensify.com/api.php?command` route and it seems like it would not be that easy to "funnel" the response through the `new.expensify.com` to avoid cors.

I think it should be possible to proxy the request like:
1. client from `new.expensify.com` makes a request to `new.expensify.com/chat-attachments/...` 
2. The server at `new.expensify.com` forwards this to `www.epxensify.com/api` 
3. `www.expensify.com` performs the requests and returns result
4. which `new.expensify.com` serves to the client

**Comment by mountiny** [comment]
> We're already using the custom header in native iOS/Android apps - no problems there, because they aren't subject to CORS checks (they can make direct requests to any URL) - only browsers run CORS checks

Ah forgot that was only on native!

Thanks for the write up! When do we plan to use the header for web as well? Do you by any chance have a link to that issue? I am just trying to get some idea for how to prioritize this and when we would technically start seeing some issues.

> I think it should be possible to proxy the request like:

I will discuss this internally, I am not that familiar with the server settings. I feel same as you, as this should be possible, but lets see

**Comment by kidroca** [comment]
@mountiny 
This is the issue handling using headers on web: 
- https://github.com/Expensify/App/issues/12603

We've already added support for Image headers in `react-native-web` 

What's remaining is to add small updates to App Image component for web, handled here:
- https://github.com/Expensify/App/pull/13036

Ofcourse, we'd also need to address the CORS issue one way or another in order to proceed

**Comment by mountiny** [comment]
@kidroca thanks, thats useful context.

Since the PR is almost ready to be merge, I take this as sooner is better than later.

**Comment by mountiny** [comment]
@bfitzexpensify seems like we might have a regression there so I would wait with payment

**Comment by mountiny** [comment]
@kidroca @mananjadhav Seems like the PR did not pass QA:


## Action Performed:
Break down in numbered steps
1. On staging, upload a PDF
2. On production, try to open the PDF

## Expected Result:
Describe what you think should've happened
PDF should load

## Actual Result:
Describe what actually happened
![Screenshot 2023-02-01 at 3 09 19 PM](https://user-images.githubusercontent.com/30609178/216187151-dabb522a-b78f-47c1-bd89-d54c40091623.png)

Would you be bale to look into this? I feel like this might have been present before our change

**Comment by mananjadhav** [comment]
I'll definitely check this later today.  But quick point, the PDF that I tested on my local was something I uploaded on production, and then tested on all the platforms locally. So I am not sure what's breaking. I'll take a quick stab at it later today.

**Comment by mountiny** [comment]
I think the breaking case was PDF uploaded on staging and then failed to load on production. Havent looked into this yet either

**Comment by kidroca** [comment]
@mountiny 
I've already posted about the exact same PDF bug earlier: https://github.com/Expensify/App/issues/14227#issuecomment-1384151477

It was actual before my PR was merged, but you can revert my PR to verify the bug would still exist

AFAIK my changes are on `staging` (only). The changes do not affect the upload process 
Since my changes are not on `prod` they can't possibly break that environment 

The PR should actually fix prod once it's deployed there

**Comment by mananjadhav** [comment]
@kidroca @mountiny Is there a reliable way to reproduce this? I just uploaded a PDF on staging and viewed on production, and vice-versa.


https://user-images.githubusercontent.com/3069065/216347169-97fd5b58-4c91-40f3-bbf6-6663ffa42082.mov

**Comment by mountiny** [comment]
> I've already posted about the exact same PDF bug earlier: https://github.com/Expensify/App/issues/14227#issuecomment-1384151477

@kidroca Hah, there is too many thread and github issues these days, sorry for missing that

@mananjadhav sometimes it not every PDF which does these I believe.

**Comment by mountiny** [comment]
@kidroca Amazing, thank you for confirming ❤️ Glad we did not revert it

**Comment by kidroca** [comment]
@mountiny 
> > I think it should be possible to proxy the request like:
> 
> I will discuss this internally, I am not that familiar with the server settings. I feel same as you, as this should be possible, but lets see

Can someone post an update regarding the CORS issue, and what's decided as a fix 
My guess is that the API server would be updated so it allows requests originating from `new.expensify` and `staging.new.expensify` and allowing the `X-Chat-Attachment-Token` header, but I need someone to confirm

**Comment by mountiny** [comment]
> My guess is that the API server would be updated so it allows requests originating from new.expensify and staging.new.expensify and allowing the X-Chat-Attachment-Token header, but I need someone to confirm

I still need to look into this, sorry for the delays

**Comment by mananjadhav** [comment]
@mountiny this is ready for the payout right? or we're going to wait for the CORS fix to close this one out?

**Comment by mountiny** [comment]
Yes it is @bfitzexpensify would you be able to pay @mananjadhav $1000 for reviewing and testing the PR? 

I would keep this open for the cors fix yet

**Comment by kidroca** [comment]
@mountiny 
Regarding [my comment](https://github.com/Expensify/App/issues/14227#issuecomment-1416052221) and the remaining CORS issue, should we open a separate ticket? 
Or maybe there's already an existing internal issue? 

I don't think I can do anything about it in the scope of the current ticket (needs backend update, can't be fixed with client side changes)

**Comment by mountiny** [comment]
Yeah I think this might be better to abstract into its own issue. Would you be bale to write down exactly what we need to allow in the servers to avoid the CORS issues? I could try to get someone from infra team involved to help with this.

If you post it here as and issue body, I can create the new issue and take it from there for clarity

**Comment by getusha** [comment]
@mountiny we want to disable the `CORS` completely from the client side?

**Comment by kidroca** [comment]
@mountiny 
Links to comments explaining the CORS issue
- https://github.com/Expensify/App/issues/14227#issuecomment-1406800255
- https://github.com/Expensify/App/issues/14227#issuecomment-1410089720

### Summary

We are removing the auth token from the URL part of an attachment/image and instead placing it in a HTTP header called `X-Chat-Attachment-Token` 
We're doing this to improve file caching (when the token is part of the URL, the URL changes often and cannot be cached) 

As a consequence (for adding this header) browsers start to make **preflight** (OPTIONS method) requests 
(This also applies to Desktop since under the hoods it's using Chromium) 

The **preflight** requests are being made because the API url (we request the image from) is from a different origin
than the place where webapp is hosted 
e.g. website is `new.expensify.com` while API is `www.expensify.com` 

Right now the OPTIONS request, seems unexpected and the API redirects with a response to an error page and we get the following error

Web:
```
[Error] Preflight response is not successful. Status code: 302
[Error] Fetch API cannot load https://www.expensify.com/chat-attachments/4198796077944938521/w_6f21f50b3af2ee93c4e6e288ba5b986858166367.jpg.1024.jpg due to access control checks.
[Error] Failed to load resource: Preflight response is not successful. Status code: 302 (w_6f21f50b3af2ee93c4e6e288ba5b986858166367.jpg.1024.jpg, line 0)
[Error] Preflight response is not successful. Status code: 302
```

Desktop:
```
Access to fetch at 'https://www.expensify.com/chat-attachments/962115596/w_e989bfe450cb8e872f52dac2d082c391f35e2e0a.jpg.1024.jpg' from origin 'http://localhost:8081' 
has been blocked by CORS policy: Response to preflight request doesn't pass access control check: Redirect is not allowed for a preflight request.
```

### What can we do 

I think this needs to be planned internally 
We can start a conversation in Slack to get a general consensus on how to deal with the CORS issue

Perhaps the simplest thing would be to update the backend to handle cross origin request coming from these origins 
- localhost
- desktop app
- staging url
- prod url 

Typically servers support a CORS configuration that can specify to allow and properly respond to request (including preflight) from these origin 

Alternatively we can avoid CORS altogether if the API is resolved from the same origin where the app is hosted
this can be a simple proxy that takes requests coming to `new.expensify.com/api` and resolves them from `www.expensify.com/api` 

Perhaps more alternative exist and someone can propose something better

**Comment by mountiny** [comment]
Thank you @kidroca, I dont think we want to get rid of CORS altogether @getusha

**Comment by mountiny** [comment]
Created a new issue [here](https://github.com/Expensify/App/issues/14991), keeping this open for payment to @mananjadhav

**Comment by kidroca** [comment]
Hey @mountiny I noticed something else while inspecting actual staging & prod environments 

`URL_API_ROOT` is `https://www.expensify.com` for both staging and prod environments, you can see the same if you put a breakpoint here:
https://github.com/Expensify/App/blob/72a2950a04df5b4a4f118a8b2d38854fc51d5ff1/src/libs/tryResolveUrlFromApiRoot.js#L22-L24

So attachment requests are always made against `https://www.expensify.com`, while the expected behavior was that
- staging would make requests against `https://staging.expensify.com`
- prod would make requests against `https://www.expensify.com`

The I saw this in `.env.staging` 
https://github.com/Expensify/App/blob/8bc6f523aa45d970d122a9f49ed18273a6987228/.env.staging#L3

`URL_API_ROOT` resolves from `EXPENSIFY_URL`, and as we can see it's `www.expensify.com` even in the staging config 

But then I've asked myself "if URL_API_ROOT is not staging.expensify.com, how come the staging app is making COMMAND requests against `staging.expensify.com` and not `www.expensify.com` 

The answer is here:
https://github.com/Expensify/App/blob/7c53f31913c0b6369852f9dcca9aab6e88a6fedf/src/libs/HttpUtils.js#L112-L114

I'm not sure what the reason for this comment
https://github.com/Expensify/App/blob/7c53f31913c0b6369852f9dcca9aab6e88a6fedf/src/libs/HttpUtils.js#L11-L12

Why should desktop and web default to staging? IMO they should default to whatever .env.staging says should be the API url

---

All that said, we see that resolving attachments from `https://www.expensify.com` works on both staging and prod, so I'm not sure if we should do something about this 

But if we do make the following a fact
- staging would make requests against `https://staging.expensify.com`
- prod would make requests against `https://www.expensify.com`

We'd need to place the same `shouldUseStagingServer` in `tryResolveUrlFromApiRoot.js`

I'm not a fan of this, because we already missed it once - it's not obvious unless someone points out that `URL_API_ROOT` is not really the address that will be used and that we should use `shouldUseStagingServer` whenever we have such logic 

I think this should be redesigned in such a way so that we import `URL_API_ROOT` or `SECURE_API_ROOT`  and these URLs are already taken care of - they point to staging when they need to

**Comment by mountiny** [comment]
@kidroca 
> Why should desktop and web default to staging? IMO they should default to whatever .env.staging says should be the API url

This means we would like web staging and web desktop to point to staging APIs by default, the toggle in settings can change it to production if we want to yes.

I am also not sure if we should actively be fixing this now as it worked like this for a while :D

I am wondering though why the .env.staging has non-staging url defined, not sure.

@kidroca Could we use the `shouldUseStagingServer` in `tryResolveUrlFromApiRoot ` as well to get around this prblem and fetch from the correct URL?

**Comment by kidroca** [comment]
> > Why should desktop and web default to staging? IMO they should default to whatever .env.staging says should be the API url
> 
> This means we would like web staging and web desktop to point to staging APIs by default, the toggle in settings can change it to production if we want to yes.

Hmm, I didn't understand the comment like that 
What I thought it meant is the default value (for web/desktop) is always staging (even if we're on prod)

Only after, seeing your comment and tracing the code I see that this is meant as the **default value for the toggle** if Onyx doesn't have any value for it at the moment 

And not meaning that by default web/desktop should be using staging api

My question though is more about, what's the problem with the other platforms - why can't we simple use CONFIG as the default value:

Instead of this:
```js
const shouldDefaultToStaging = _.contains([CONST.PLATFORM.WEB, CONST.PLATFORM.DESKTOP], getPlatform());
```

Have this:
```js
const defaultStagingServerToggleState = CONFIG.IS_IN_STAGING;
```

---

> @kidroca Could we use the `shouldUseStagingServer` in `tryResolveUrlFromApiRoot ` as well to get around this prblem and fetch from the correct URL?

Inspecting the code, I see the toggle `shouldUseStagingServer` is something that only affects STAGING where you can toggle between staging api and prod api from settings 
This is very similar to a platform check - it's a inline ENV check that alters logic, and as such I think it should be captured in a separate file, loaded only when ENV is staging - the same thing we do for platform specific code 

We could apply a workaround and use `shouldUseStagingServer` in `tryResolveUrlFromApiRoot`, but I think such code leads to problems. I already missed it entirely here (same as everyone involved in the PR) 

IMO instead of imperative checks testing for certain conditions, the logic used here (and other URL dealing places) should be setup in a declarative way where it just says 

```js
const apiRoot = getApiRoot(request);
```

Where `getApiRoot` is making the necessary checking and returns the staging/prod or secure URLs based on the request 

This also solves another problem (at least for me), 
I can't trace what logic would ever set the `shouldUseSecure` to `true` - I've digged through Apps code and followed usages and just can't find anything that would set that value to `true` 

The flag value is probably based on the command or something like that, this can be made more transparent if the conditions for a secure URL are declared in the same place e.g. use secure API for a list of command names

---

I think we should discuss this on Slack and agree whether we should 
- patch this up with another `shouldUseStagingServer` check
- do nothing
- refactor and create a function like `getApiRoot` then update usages

I also would like to either continue in a new ticket or increase the scope/payment for the current one
I've applied and accepted a job in Upwork and haven't tracked any of the work done here under my hourly contract

**Comment by mountiny** [comment]
> My question though is more about, what's the problem with the other platforms - why can't we simple use CONFIG as the default value:

Because native apps are only build once using the staging config, then after QA is done, the staging build is pushed to production, but there is no new compilation occurring. Hence for native the Config on production is still staging so that is why we need a different approach for those.

> This also solves another problem (at least for me),
I can't trace what logic would ever set the shouldUseSecure to true - I've digged through Apps code and followed usages and just can't find anything that would set that value to true

> The flag value is probably based on the command or something like that, this can be made more transparent if the conditions for a secure URL are declared in the same place e.g. use secure API for a list of command names

Yeah I am not sure straight from top of my head which commands use it in NewDot its possible we are deprecating this too and we are moving the secure commands to normal API, it is a relict and it will be deprecated down the line I believe. 

> I think we should discuss this on Slack and agree whether we should

Would you start that discussion? I think we would like to make sure that if the staging toggle is turned to staging we should resolve the attachments form there too. Declarative way makes sense to me, this part of the code is not super clear because of the platform dependency.

> I also would like to either continue in a new ticket or increase the scope/payment for the current one
I've applied and accepted a job in Upwork and haven't tracked any of the work done here under my hourly contract

We can create a new ticket after the discussion and after we decide what is needed, does that work for you?

**Comment by mountiny** [comment]
@bfitzexpensify would you be able to also pay out @kidroca for this job?

**Comment by mountiny** [comment]
@bfitzexpensify Bumping this one!

**Comment by mountiny** [comment]
Created a new ticket for the issue mentioned [here](https://github.com/Expensify/App/issues/15087)

**Comment by kidroca** [comment]
@mountiny 
Thanks for the explanations man, Kudos!

Saw the new ticket 
I'll start discussion on Slack and propose to have `getApiRoot()`, but if we decide something else in the end I'll work with that

**Comment by spreadmycode** [comment]
**Solution:**
The following condition should meet for this issue.
Backend and staging.new.expensify.com should be running on the same machine (i.e VPS or EC2 AWS).

1. Backend receive the uploading request from https://new.expensify.com/ and store the files in staging.new.expensify.com 's public folder. The reason why uploaded files should be stored in public folder of staging.new.expensify.com is to allow anyone can see (only read) these uploaded images (or digital assets).

2. Uploaded digital asset's access link (https://staging.new.expensify.com/1.png) will be stored in database as a reference to attachment for the specific chat item. You can check this public URL can access the uploaded asset to load on browser.
If needs, we should store these files on the specific folder which indicates the unique chat room or some thing like that to avoid overwriting or conflicts. It would be much better for us to create Git ignored folder in public of staging.new.expensify.com base path.

3. Finally, all assets can be fetched from staging.new.expensify.com, not new.expensify.com.

**Comment by mountiny** [comment]
@spreadmycode thanks for the proposal, can you check issues which have Help Wanted label though, there you can be assigned to work on them, this one is being closed.

I think @bfitzexpensify has paid out what was needed here so closing
